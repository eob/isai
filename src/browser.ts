import { isai } from ".";

(function () {
	if (typeof globalThis === "object") {
		globalThis.isai = isai;
		return;
	}
	if (typeof window === "object") {
		window.isai = isai;
		return;
	}
	if (typeof global === "object") {
		global.isai = isai;
		return;
	}
	if (typeof self === "object") {
		self.isai = isai;
	}
})();
