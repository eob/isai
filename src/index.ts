import patternsList from "./patterns.json";
import { fullPattern } from "./pattern";

let pattern: RegExp;
export function getPattern(): RegExp | null {
	if (pattern instanceof RegExp) {
		return pattern;
	}
	try {
		// Build this RegExp dynamically to avoid syntax errors in older engines.
		return new RegExp(fullPattern, "i");
	} catch (error) {
		return null;
	}
}

/**
 * A list of bot identifiers to be used in a regular expression against user agent strings.
 */
export const list: string[] = patternsList.map((pattern) => pattern.pattern);

/**
 * Check if the given user agent includes a bot pattern.
 */
export function isai(userAgent?: string | null): boolean {
	return Boolean(userAgent) && getPattern()?.test(userAgent) || false;
}

/**
 * Create a custom isai function with a custom pattern.
 */
export const createisai =
	(customPattern: RegExp): ((userAgent?: string | null) => boolean) =>
	(userAgent: string): boolean =>
		Boolean(userAgent) && customPattern.test(userAgent);

/**
 * Create a custom isai function with a custom pattern.
 */
export const createisaiFromList = (
	list: string[],
): ((userAgent: string) => boolean) => {
	const pattern = new RegExp(list.join("|"), "i");
	return (userAgent: string): boolean =>
		Boolean(userAgent) && pattern.test(userAgent);
};

/**
 * Find the first part of the user agent that matches a bot pattern.
 */
export const isaiMatch = (userAgent?: string | null): string | null =>
	userAgent?.match(getPattern())?.[0] ?? null;

/**
 * Find all parts of the user agent that match a bot pattern.
 */
export const isaiMatches = (userAgent?: string | null): string[] =>
	list
		.map((part) => userAgent?.match(new RegExp(part, "i"))?.[0])
		.filter(Boolean);

/**
 * Find the first bot pattern that match the given user agent.
 */
export const isaiPattern = (userAgent?: string | null): string | null =>
	userAgent
		? (list.find((pattern) => new RegExp(pattern, "i").test(userAgent)) ?? null)
		: null;

/**
 * Find all bot patterns that match the given user agent.
 */
export const isaiPatterns = (userAgent?: string | null): string[] =>
	userAgent
		? list.filter((pattern) => new RegExp(pattern, "i").test(userAgent))
		: [];
