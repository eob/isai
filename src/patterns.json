[
  {
    "pattern": "GoogleOther", 
    "confidence": 0.5,
    "reasonsFor": [
      "Google Notebook LLM (https://notebooklm.google/) uses this user agent string."
    ],
    "reasonsAgainst": [
      "GoogleOther is known to be a general-purpose agent string for internal Google projects."
    ]
  },
  {
    "pattern": "https://openai.com/gptbot", 
    "confidence": 1.0,
    "reasonsFor": [
      "Published user agent substring by OpenAI"
    ],
    "reasonsAgainst": [
    ]
  },
  {
    "pattern": "https://openai.com/searchbot", 
    "confidence": 1.0,
    "reasonsFor": [
      "Published user agent substring by OpenAI"
    ],
    "reasonsAgainst": [
    ]
  },
  {
    "pattern": "https://openai.com/bot", 
    "confidence": 1.0,
    "reasonsFor": [
      "Published user agent substring by OpenAI"
    ],
    "reasonsAgainst": [
    ]
  }
]
