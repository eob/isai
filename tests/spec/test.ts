import {
	getPattern,
	list,
	isai,
	isaiMatch,
	isaiMatches,
	isaiPattern,
	isaiPatterns,
	createisai,
	createisaiFromList,
} from "../../src";
import { fullPattern } from "../../src/pattern";
import { crawlers, browsers } from "../../fixtures";
let isaiInstance: any;

const AI_USER_AGENT_EXAMPLE =
	"Mozilla/5.0 AppleWebKit/537.36 (KHTML, like Gecko); compatible; OAI-SearchBot/1.0; +https://openai.com/searchbot";
const BROWSER_USER_AGENT_EXAMPLE =
	"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91 Safari/537.36";

describe("isai", () => {
	describe("features", () => {
		test("pattern: pattern is a regex", () => {
			expect(getPattern()).toBeInstanceOf(RegExp);
		});
		test("list: list is an array", () => {
			expect(list).toBeInstanceOf(Array);
			expect(list.every((item) => typeof item === "string")).toBe(true);
		});
		test("isai: bot user agent string is recognised as an AI", () => {
			expect(isai(AI_USER_AGENT_EXAMPLE)).toBe(true);
		});
		test("isaiMatch: find pattern in bot user agent string", () => {
			expect(isaiMatch(AI_USER_AGENT_EXAMPLE)).toBe("openai.com/searchbot");
		});
		test("isaiMatches: find all patterns in bot user agent string", () => {
			expect(isaiMatches(AI_USER_AGENT_EXAMPLE)).toContain("openai.com/searchbot");
			expect(isaiMatches(AI_USER_AGENT_EXAMPLE)).toHaveLength(1);
		});
		test("isaiPattern: find first pattern in bot user agent string", () => {
			expect(isaiPattern(AI_USER_AGENT_EXAMPLE)).toBe(
				"openai.*bot",
			);
		});
		test("isaiPatterns: find all patterns in bot user agent string", () => {
			expect(isaiPatterns(AI_USER_AGENT_EXAMPLE)).toContain(
				"openai.*bot",
			);
			expect(isaiPatterns(AI_USER_AGENT_EXAMPLE)).toHaveLength(1);
		});
		test("createisai: create custom isai function with custom pattern", () => {
			const customisai = createisai(/bot/i);
			expect(customisai(AI_USER_AGENT_EXAMPLE)).toBe(true);
		});
		test("createisaiFromList: create custom isai function with custom pattern", () => {
			const ToRemoveStrings: string[] = [
				"openai.*bot"
			];
			const patternsToRemove: Set<string> = new Set(
				ToRemoveStrings.map(isaiMatches).flat(),
			);
			expect(patternsToRemove.size).toBeGreaterThan(0);
			const list2 = list.filter(
				(record: string): boolean => patternsToRemove.has(record) === false,
			);
			expect(list2.length).toBeLessThan(list.length);
			const isai2 = createisaiFromList(list2);
			const ua = "https://openai.com/gptbot"
			expect(isai(ua)).toBe(true);
			expect(isai2(ua)).toBe(false);
		});
		test.each([null, undefined, ""])(
			"all functions can accept %p",
			(value: string | null | undefined) => {
				expect(isai(value)).toBe(false);
				expect(isaiMatch(value)).toBe(null);
				expect(isaiMatches(value)).toEqual([]);
				expect(isaiPattern(value)).toBe(null);
				expect(isaiPatterns(value)).toEqual([]);
			},
		);
	});

	describe("regex fallback", () => {
		beforeAll(async () => {
			jest
				.spyOn(globalThis, "RegExp")
				.mockImplementation((pattern, flags): RegExp => {
					if ((pattern as string).includes?.("?<!")) {
						throw new Error("Invalid regex");
					}
					return new RegExp(pattern, flags);
				});
			const mdl = await import("../../index.js");
			if (!mdl) {
				throw new Error("Module not found");
			}
			isaiInstance = mdl.isai as ReturnType<typeof createisai>;
		});
		afterAll(() => {
			jest.restoreAllMocks();
		});
		test("fallback does not detect browser as bot", () => {
			expect(isaiInstance(BROWSER_USER_AGENT_EXAMPLE)).toBe(false);
		});
	});

	describe("fixtures", () => {
		test(`✔︎ ${crawlers.length} user agent string should be recognised as crawler`, () => {
			let successCount = 0;
			let misidentifiedStrings: string[] = [];
			crawlers.forEach((crawler) => {
				if (isai(crawler)) {
					successCount++;
				} else {
					misidentifiedStrings.push(crawler);
				}
			});
			expect(misidentifiedStrings).toEqual([]);
			expect(successCount).toBe(crawlers.length);
		});
		test(`✘ ${browsers.length} user agent string should not be recognised as an AI`, () => {
			let successCount = 0;
			let misidentifiedStrings: string[] = [];
			browsers.forEach((browser) => {
				if (isai(browser)) {
					misidentifiedStrings.push(browser);
				} else {
					successCount++;
				}
			});
			expect(misidentifiedStrings).toEqual([]);
			expect(successCount).toBe(browsers.length);
		});
	});

	describe("module interface", () => {
		test("interface is as expected", async () => {
			const types = Object.entries(await import("../../src/index")).map(
				([key, value]) => [key, value.constructor.name] as [string, string],
			);
			expect(types).toMatchSnapshot();
		});
		test("regular expressions exports are as expected", () => {
			expect(new RegExp(fullPattern, "i").toString()).toBe(
				getPattern()?.toString(),
			);
		});
	});
});
