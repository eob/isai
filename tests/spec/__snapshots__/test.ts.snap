// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`isai module interface interface is as expected 1`] = `
[
  [
    "list",
    "Array",
  ],
  [
    "isaiNaive",
    "Function",
  ],
  [
    "createisai",
    "Function",
  ],
  [
    "createisaiFromList",
    "Function",
  ],
  [
    "isaiMatch",
    "Function",
  ],
  [
    "isaiMatches",
    "Function",
  ],
  [
    "isaiPattern",
    "Function",
  ],
  [
    "isaiPatterns",
    "Function",
  ],
  [
    "getPattern",
    "Function",
  ],
  [
    "isai",
    "Function",
  ],
]
`;
